tests/phpunit/tests/blocks/wpBlockType.php

```diff
@@ -460,4 +460,190 @@ public function data_block_version() {
+					return $this->mock_variation_callback();
+				},
+			)
+		);
+
+		$this->assertSame( false, $callback_called, 'The callback should not be called before the variations are accessed.' );
+		$block_type->variations; // access the variations.
+		$this->assertSame( true, $callback_called, 'The callback should be called when the variations are accessed.' );
+	}
+
+	/**
+	 * @ticket 59969
+	 * @covers WP_Block_Type::get_variations
+	 */
+	public function test_variations_precedence_over_callback_post_registration() {
+		$test_variations = array( 'name' => 'test1' );
+		$callback_called = false;
+
+		$block_type             = new WP_Block_Type(
+			'test/block',
+			array(
+				'title'              => 'Test title',
+				'variation_callback' => function () use ( &$callback_called ) {
+					$callback_called = true;
+					return $this->mock_variation_callback();
+				},
+			)
+		);
+		$block_type->variations = $test_variations;
+
+		// If the variations are defined after registration but before first access, the callback should not override it.
+		$this->assertSameSets( $test_variations, $block_type->get_variations(), 'Variations are same as variations set' );
+		$this->assertSame( false, $callback_called, 'The callback was never called.' );
+	}
+
+	/**
+	 * @ticket 59969
+	 * @covers WP_Block_Type::get_variations
+	 */
+	public function test_variations_callback_happens_only_once() {
+		$callback_count = 0;
+
+		$block_type = new WP_Block_Type(
+			'test/block',
+			array(
+				'title'              => 'Test title',
+				'variation_callback' => function () use ( &$callback_count ) {
+					$callback_count++;
+					return $this->mock_variation_callback();
+				},
+			)
+		);
+
+		$this->assertSame( 0, $callback_count, 'The callback should not be called before the variations are accessed.' );
+		$block_type->get_variations(); // access the variations.
+		$this->assertSame( 1, $callback_count, 'The callback should be called when the variations are accessed.' );
+		$block_type->get_variations(); // access the variations again.
+		$this->assertSame( 1, $callback_count, 'The callback should not be called again.' );
+	}
+
+	/**
+	 * @ticket 59969
+	 */
+	public function test_get_block_type_variations_filter_with_variation_callback() {
+		// Filter will override the variations obtained from the callback.
+		add_filter( 'get_block_type_variations', array( $this, 'filter_test_variations' ), 10, 2 );
+		$expected_variations = array( array( 'name' => 'test1' ) );
+
+		$callback_called = false;
+		$block_type      = new WP_Block_Type(
+			'test/block',
+			array(
+				'title'              => 'Test title',
+				'variation_callback' => function () use ( &$callback_called ) {
+					$callback_called = true;
+					return $this->mock_variation_callback();
+				},
+			)
+		);
+
+		$obtained_variations = $block_type->variations; // access the variations.
+		remove_filter( 'get_block_type_variations', array( $this, 'filter_test_variations' ), 10 );
```

Comment:
Hooks get reset in `WP_UnitTestCase_Base::set_up` that runs before each test ([link](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/includes/abstract-testcase.php#L111-L113)), so no need to clean up filters manually in tests.

🔢 Applies here and throughout.

Reply:
Done.